File: src/main/java/com/leetcode/aprilchallenge/MinStack.java
package com.leetcode.aprilchallenge;

import java.util.PriorityQueue;
import java.util.Stack;

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 *
 * @author wezhyn
 * @since 04.11.2020
 */
public class MinStack {


    private PriorityQueue<Integer> priorityQueue=new PriorityQueue<>();
    private Stack<Integer> stack=new Stack<>();
    /**
     * initialize your data structure here.
     */
    public MinStack() {
    }

    public void push(int x) {
        stack.push(x);
        priorityQueue.offer(x);
    }

    public void pop() {
        final Integer pop=stack.pop();
        if (pop!=null) {
            priorityQueue.remove(pop);
        }
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return priorityQueue.peek();
    }

}

File: leetcode-kotlin/src/main/kotlin/com/wezhyn/learn/lcof/MinStack.kt
package com.wezhyn.learn.lcof

import java.util.*

/**
 *
 * @author wezhyn
 * @since 06.30.2020
 *
 */
class MinStack {

    private val stack = LinkedList<Int>()
    private val minStack = LinkedList<Int>()

    fun push(x: Int) {
        stack.push(x)
        val peek = minStack.peek()
        if (x < peek) {
            minStack.push(x)
        } else {
            minStack.push(peek)
        }
    }

    fun pop() {
        stack.poll()
        minStack.poll()
    }

    fun top(): Int {
        return stack.peek()
    }

    fun getMin(): Int {
        return minStack.peek()
    }
}